File: Partition Array for Maximum Sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        n = len(arr)
        max_sum = [0] * (n + 1)

        for i in range(1, n + 1):
            mpn = float("-inf")
            for mpn_count in range(1, min(i, k) + 1):
                mpn = max(mpn, arr[i - mpn_count])
                max_sum[i] = max(max_sum[i], max_sum[i - mpn_count] + mpn * mpn_count)
        
        return max_sum[n]

File: Partition Array for Maximum Sum.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number}
 */
var maxSumAfterPartitioning = function(arr, k) {
    const n = arr.length;
    const maxSum = new Array(n + 1).fill(0);

    for (let i = 1; i <= n; i++) {
        let mpn = Number.MIN_SAFE_INTEGER;
        for (let mpnCount = 1; mpnCount <= Math.min(i, k); mpnCount++) {
            mpn = Math.max(mpn, arr[i - mpnCount]);
            maxSum[i] = Math.max(maxSum[i], maxSum[i - mpnCount] + mpn * mpnCount);
        }
    }

    return maxSum[n];    
};
